pkg/controller/controllerutil/controllerutil.go

```diff
@@ -334,6 +336,47 @@ func CreateOrPatch(ctx context.Context, c client.Client, obj client.Object, f Mu
 	return result, nil
 }
 
+// CreateIfNotExist creates the given object in the Kubernetes
+// cluster. The object's desired state must be reconciled with the existing
+// state inside the passed in callback MutateFn.
+//
+// It returns the executed operation and an error.
+func CreateIfNotExist(ctx context.Context, c client.Client, obj client.Object, f MutateFn) (OperationResult, error) {
+	key := client.ObjectKeyFromObject(obj)
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return OperationResultNone, err
+		}
+		if err := mutate(f, key, obj); err != nil {
+			return OperationResultNone, err
+		}
+		if err := c.Create(ctx, obj); err != nil {
+			return OperationResultNone, err
+		}
+		return OperationResultCreated, nil
+	}
+	return OperationResultNone, nil
+}
+
+// DeleteIfExist deletes the given object in the Kubernetes
+// cluster.
+//
+// It returns the executed operation and an error.
+func DeleteIfExist(ctx context.Context, c client.Client, obj client.Object, f MutateFn) (OperationResult, error) {
+	key := client.ObjectKeyFromObject(obj)
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return OperationResultNone, err
+		}
+		if err := c.Delete(ctx, obj); err != nil {
+			return OperationResultNone, err
```

Comment:
Generally in code like this you want to allow a 404 response to still be considered a success, helps with concurrency conditions.

Reply:
@coderanger So if I got it right, irrespective of the error has occurred or not the response should be OperationResultDeleted? right?

Reply:
I would probably return `OperationResultNone, nil` if the delete returns a 404? This code didn't do a delete but someone else did.

Reply:
Well if the object did not exist(404) it will never enter the if statement and default returns `OperationResultNone, nil` but if it does exist and delete could not process the action hence returned the error.

Reply:
That would not have returned a 404 though, a 404 specifically means "object not found".

Reply:
It's possible for `c.Get` to hand you an object and `c.Delete` to return a 404. The reason is that `c.Get` returns an object from the cache (which might be outdated), and `c.Delete` issues the request directly to the apiserver. And even if `c.Get` read from the apiserver directly, it would be possible for the apiserver to receive and handle a deletion request from another client between the `c.Get` and `c.Delete` calls.

I think you can simplify to this instead:
```go
if err := c.Delete(ctx, obj); client.IgnoreNotFound(err) != nil {
	return OperationResultNone, err
}
return OperationResultDeleted, nil
```

But given that calling code would take up just as many lines as this wrapper, I wonder if there's any actual value to the `DeleteIfExists` wrapper function?

Reply:
Given the eventual destiny of CreateOrUpdate is be removed in favor of Server Side Apply (some day, don't anyone go panic this will be soon) I could see an argument we shouldn't really add more of these helpers in general. But maybe having a "common patterns" doc to help folks coming up to speed with using the API?

Reply:
(OTOH treating this file as "executable documentation" wouldn't be the worst)

Reply:
@joelanford I tend to agree. But there are few libraries that keep adding these types of methods.

Reply:
@jmrodri A lot of folks struggle to understand writing convergent code and try to shoehorn procedural patterns into operators where they don't belong. Very understandable, but that doesn't mean we should encourage those patterns. We should instead improve our resources to teach people convergent patterns :)

Reply:
If we update the deleteIfExits to
```
if err := c.Delete(ctx, obj); client.IgnoreNotFound(err) != nil {
	return OperationResultNone, err
}
return OperationResultDeleted, nil
```
It will return `OperationResultDeleted` even if the object doesn't exist. If that's okay? I will go ahead and update the helper.

Reply:
I'm trying to think of a scenario where I would call this and need to handle `OperationResultDeleted, nil` differently from `OperationResultNone, nil`.

In either case, the thing I want deleted is not present after this returns a non-nil error.

Reply:
@joelanford The usual case is emitting Events for debugging or analysis purposes.

Reply:
@coderanger Good point. Slight tweak then:

```go
if err := c.Delete(ctx, obj); err != nil {
	if apierrors.IsNotFound(err) {
		return OperationResultNone, nil
	}
	return OperationResultNone, err
}
return OperationResultDeleted, nil
```

Reply:
@joelanford @coderanger thanks for the feedback. Updated the helper.

Reply:
As @joelanford previously mentioned, I'm thinking of the benefit this would add to the generic `client.Delete`. We are additionally wrapping `IsNotFound` error and returning `OperationResultNone`. Regarding handling of `OperationResultDeleted` and `OperationResultNone`, if we want to find if the object exists in the cluster or not, shouldn't the right method be to send the request again and use `IsNotFound` error directly. It seems counterintuitive that if someone wants to delete an object either way, they would be handing the case where object wasn't present in the cluster differently. Or is there any scenrio I am missing?

Reply:
@varshaprasad96 @coderanger @varshaprasad96 @jmrodri 
Any further discussion on this do we need any amends?

Reply:
@joelanford @coderanger @jmrodri - would like to know your opinion. I am still a bit doubtful on whether having this helper would be useful. If not, we could have this documented in FAQ instead of creating a separate wrapper.

Reply:
So what I'm reading is that call to `Delete` would look like this:
```go
if err := c.Delete(ctx, obj); err != nil {
	if apierrors.IsNotFound(err) {
		// handle not found
	}
	// handle error
}
// otherwise deleted
```

and a call `DeleteIfExists` would look like:

```go
if _, err := controllerutil.DeleteIfExist(ctx, c, obj); err != nil {
    // handle error
}
// otherwise deleted
```

I still see value in not having to handle the `IsNotFound` case each time. But I can see that it's not that much more code. I still see the value of `DeleteIfExists`, that's just my opinion.